externref: implement stack map-based garbage collection

fitzgen opened the pull request:
For host VM code, we use plain reference counting, where cloning increments the reference count, and dropping decrements it. We can avoid many of the on-stack increment/decrement operations that typically plague the performance of reference counting via Rust's ownership and borrowing system. Moving a `VMExternRef` avoids mutating its reference count, and borrowing it either avoids the reference count increment or delays it until if/when the `VMExternRef` is cloned.

When passing a `VMExternRef` into compiled Wasm code, we don't want to do reference count mutations for every compiled `local.{get,set}`, nor for every function call. Therefore, we use a variation of **deferred reference counting**, where we only mutate reference counts when storing `VMExternRef`s somewhere that outlives the activation: into a global or table. Simultaneously, we over-approximate the set of `VMExternRef`s that are inside Wasm function activations. Periodically, we walk the stack at GC safe points, and use stack map information to precisely identify the set of `VMExternRef`s inside Wasm activations. Then we take the difference between this precise set and our over-approximation, and decrement the reference count for each of the `VMExternRef`s that are in our over-approximation but not in the precise set. Finally, the over-approximation is replaced with the precise set.

The `VMExternRefActivationsTable` implements the over-approximized set of `VMExternRef`s referenced by Wasm activations. Calling a Wasm function and passing it a `VMExternRef` moves the `VMExternRef` into the table, and the compiled Wasm function logically "borrows" the `VMExternRef` from the table. Similarly, `global.get` and `table.get` operations clone the gotten
`VMExternRef` into the `VMExternRefActivationsTable` and then "borrow" the reference out of the table.

When a `VMExternRef` is returned to host code from a Wasm function, the host increments the reference count (because the reference is logically "borrowed" from the `VMExternRefActivationsTable` and the reference count from the table will be dropped at the next GC).

For more general information on deferred reference counting, see *An Examination of Deferred Reference Counting and Cycle Detection* by Quinane: https://openresearch-repository.anu.edu.au/bitstream/1885/42030/2/hon-thesis.pdf

cc #929

Fixes #1804

Depends on https://github.com/rust-lang/backtrace-rs/pull/341

bjorn3 commented on crates/environ/src/cranelift.rs:
```diff
@@ -204,6 +205,27 @@ impl binemit::TrapSink for TrapSink {
+
+impl StackMapSink {
+    fn finish(mut self) -> Vec<StackMapInformation> {
+        self.infos.sort_by_key(|info| info.code_offset);
```
```suggestion
        self.infos.sort_unstable_by_key(|info| info.code_offset);
```

bjorn3 commented on crates/runtime/src/externref.rs:
```diff
@@ -307,35 +337,73 @@ impl VMExternRef {
+    // pub unsafe fn from_raw(ptr: *mut u8) -> Self {
+    //     debug_assert!(!ptr.is_null());
+    //     VMExternRef(NonNull::new_unchecked(ptr).cast())
+    // }
```
Reminder to uncomment or remove.

bjorn3 commented on crates/runtime/src/externref.rs:
```diff
@@ -393,6 +461,559 @@ impl Deref for VMExternRef {
+
+#[derive(Debug)]
+struct ModuleStackMaps {
+    /// The range of PCs that this module covers. Different modules should
```
```suggestion
    /// The range of PCs that this module covers. Different modules must
```

bjorn3 commented on crates/runtime/src/externref.rs:
Maybe panic in the drop implementation instead and add an unsafe `finish` function?

bjorn3 commented on crates/wasmtime/src/runtime.rs:
```diff
@@ -193,10 +194,15 @@ impl Config {
             .operator_config
             .enable_reference_types = enable;
-        // The reference types proposal depends on the bulk memory proposal
+        self.flags
+            .set("enable_safepoints", if enable { "true" } else { "false" })
+            .unwrap();
```
```suggestion
        if enable {
            self.flags.enable("enable_safepoints").unwrap();
        }
```

fitzgen replied:
This doesn't work for disabling reference types, and there is no `disable` function to go along with `enable` either.

bjorn3 commented on crates/wasmtime/src/trampoline/func.rs:
```diff
@@ -207,7 +209,14 @@ pub fn create_handle_with_function(
+        if ft.params().iter().any(|p| *p == ValType::ExternRef)
+            || ft.results().iter().any(|r| *r == ValType::ExternRef)
+        {
+            flag_builder.set("enable_safepoints", "true").unwrap();
```
```suggestion
            flag_builder.enable("enable_safepoints").unwrap();
```

tschneidereit commented:
Very exciting to see this! \o/

One question: does this have any implications for deterministic behavior? ISTM that as long as we don't have weak references anywhere in the system it probably doesn't, right?

alexcrichton commented on crates/runtime/src/externref.rs:
```diff
@@ -307,35 +337,73 @@ impl VMExternRef {
+    ///  `clone_from_raw` is called.
+    pub fn as_raw(&self) -> *mut u8 {
         let ptr = self.0.cast::<u8>().as_ptr();
         mem::forget(self);
```
I think this line can now be removed

alexcrichton commented on crates/runtime/src/externref.rs:
```diff
@@ -393,6 +461,559 @@ impl Deref for VMExternRef {
+
+/// A registry of stack maps for currently active Wasm modules.
+#[derive(Default)]
+pub struct StackMapRegistry {
```
We've got a lot of various registries floating around wasmtime, would it be possible to insert this into an existing registry somehow? For example most of the code here looks like it's copying what we do for frame information.

alexcrichton commented on crates/runtime/src/instance.rs:
```diff
@@ -74,6 +75,19 @@ pub(crate) struct Instance {
+    ///
+    /// The `vmctx` also holds a raw pointer to the table and relies on this
+    /// member to keep it alive.
+    pub(crate) externref_activations_table: Rc<VMExternRefActivationsTable>,
```
FWIW I've been trying to clean this up over time where `Instance` doesn't do much memory management of its internals. Is it possible to package up this and the stack map registry elsewhere? I'm not really sure entirely what that would look like but I think it would be best if we could keep `Instance` having a pretty "raw" feeling rather than holding onto various items.

(e.g. shoving this into `Store` and passing it through for when we call wasm)

alexcrichton replied:
Reading over this, I also don't think that this needs to hold onto this information? It looks like the store can hold onto it for all instances?

alexcrichton replied:
Also, to keep adding to this, I'd like to get around to cleaning up the interrupts part eventually too

fitzgen replied:
This runs into the weirdness at the `wasmtime` and `wasmtime-runtime` boundary again: this is inside `wasmtime-runtime` and so it has "no" knowledge of `Store` and whether anything else is holding the table alive for it.

Would you prefer that `InstanceHandle::new` took a `*mut VMExternRefActivationsTable` instead of an `Rc` and we added this to the safetry invariants required to be maintained for `InstanceHandle::new`?

alexcrichton commented on crates/wasmtime/src/module.rs:
```diff
@@ -82,6 +83,7 @@ struct ModuleInner {
     store: Store,
     compiled: CompiledModule,
     frame_info_registration: Mutex<Option<Option<Arc<GlobalFrameInfoRegistration>>>>,
+    stack_map_registration: Mutex<Option<Option<Arc<StackMapRegistration>>>>,
```
Sort of continuing my comment from earlier, but this is an example where it would be great to not have this duplication. Ideally there'd be one "register stuff" call, although I'm not sure yet if this is fundamentally required to be two separate steps.

alexcrichton commented on crates/wasmtime/src/ref.rs:
```diff
@@ -36,6 +36,11 @@ impl ExternRef {
         &*self.inner
```
Looks like this file become executable?

fitzgen replied:
I really need to figure out why the heck this is happening...

alexcrichton commented on crates/wasmtime/src/runtime.rs:
```diff
@@ -690,6 +696,7 @@ pub enum ProfilingStrategy {
 #[derive(Default, Clone)]
 pub struct Engine {
     config: Arc<Config>,
+    stack_map_registry: Arc<StackMapRegistry>,
```
I think I may be missing something, but how come this is stored in an `Engine`? (vs a `Store`)

fitzgen replied:
It was unnecessary, and is now removed.

fitzgen commented:
> One question: does this have any implications for deterministic behavior? ISTM that as long as we don't have weak references anywhere in the system it probably doesn't, right?

This has deterministic (albeit perhaps surprising if you don't know the implementation) behavior. We only GC when either the embedder calls `Store::gc` or when the `VMExternRefActivationsTable` is full (which happens after passing in N `externref`s to Wasm).

FWIW. `VMExternRef` doesn't currently support weak references. Shouldn't be too hard to add if we find we need it. It wouldn't change the determinism of GC though.

fitzgen commented:
> Very nice! Before I dig too closely into the internal details I wanted to get a grasp of how this is all organized and such. The main thoughts I have is that this is adding a new registry and a lot of separate points where we pass around registries and such. I'm hoping we can perhaps unify all these with existing registries and such we have? Internally it feels better if we don't have an arc and rwlock per-thing that we need to keep track of.

I originally did add this to the existing global frame info, but after our discussions about not wanting implicit global context, I moved it out to its own registry.

> One other thing I'm remembering now which I think would be good to happen here, I think this should include an implementation of `WasmTy for ExternRef`. We'll want to enable closures with `Func` that take `ExternRef` as arguments and such, and it'd be cool to see what the monomorphize logic looks like for functions that take a number of `ExternRef` or produce them.

Yeah, I have this on my TODO list for a follow up PR. Felt like this was big enough as is.

> Finally I wanted to write some thoughts about the GC aspect. It feels a bit weird to me that we're using reference counting but still have to have explicit GC points. To make sure I understand, GC only happens right now automatically when you call into a function, right? Or are there other auto-inserted points that GC happens? At a minimum I think we need to make sure that all long-running code is eventually GC'd, so I think both entry and exit needs GC'ing (calling into a host and returning back to wasm may already do the GC, I likely missed it!).
> 
> I wanted to dig a bit more into the decision though to use deferred reference counting rather than explicit. Is it possible to somewhat easily get performance number comparisons? For example do we have a handle on what we predict the overhead will be? I'd imagine there are possible optimizations where local.get/local.set don't do reference counting but calling a function does.

Regarding when GC happens see my reply to Till: https://github.com/bytecodealliance/wasmtime/pull/1832#issuecomment-640756824

For long-running Wasm, in the absence of explicit `Store::gc` calls, GC will happen as the `VMExternRefActivationsTable` fills up. I don't think it makes sense to GC on entry/exit from every single host<-->wasm function boundary, because it will be too expensive for simple getter/setter host functions that you expect to generally be fast. It might makes sense for the embedder to call `Store::gc` after the "main" call into Wasm (e.g. after servicing an HTTP request, but not after each wasm -> host call to read an HTTP header; however in this case there should be zero Wasm frames on the stack, so we could provide an unsafe method to just clear the `VMExternRefActivationsTable` rather than even bothering to walk the stack and do the GC).

Regarding the decision to go with deferred reference counting, let's look at the alternative: we would have to do explicit reference count increments and decrements for references inside Wasm frames, rather than deferring them. First, to get this working at all, this would require extending `cranelift-wasm` to add ref count increments every time we pushed a reference onto the spec-defined Wasm execution stack (e.g. every `local.get`), and ref count decrements every time it is popped off the Wasm execution stack (e.g. every `local.set`). But doing this many reference counting operations is *the* thing kills reference counting's throughput. So to alleviate that we could either do deferred reference counting, or we could add a static analysis to `cranelift-wasm` that sums the ref count increments and decrements in a basic block and does only the final effect once (hand waving a bit, would need to think about this a bit more). Then, if coalescing ref count operations within a basic block isn't enough, this static analysis could perhaps further be generalized across basic blocks (really hand waving now).

Note that we can't only do reference counting at function call boundaries because we need to handle the case where a Wasm function takes a reference and drops it. We need to decrement the reference count at the drop site, which is what instrumenting the spec stack pops does.

So that is a bunch of infrastructural work on Cranelift to get something that will perform worse but at least works, followed by more work to start moving towards fewer ref count operations (but never getting as few as deferred reference counting: zero, albeit with occasional stack walking pauses).

On the other hand, we already have stack maps produced by Cranelift, which are the necessary bits required for implementing deferred reference counting. We don't need to build any Cranelift infrastructure *and* Wasmtime's integration, just the latter. However, yes, this does come with occasional stack-walking pauses and reliance on `libunwind` or some other way of walking the stack (but also we already require it for unwinding host functions after traps).

So that was basically the calculus: an easier path to getting a better implementation (or at least an easier path to getting a good implementation, if we are comparing against the hypothetical best version of the ref counting coalescing static analysis).

Unfortunately, I don't have benchmarks or performance numbers. It would be hard to get this information without implementing both approaches. And it is hard to know how many reference counting operations we could coalesce with the hypothetical static analysis. But I don't really have any doubt that naively doing all the increments and decrements for on-stack references would be quite slow: this is Well Known(tm) in the memory management world.

tschneidereit commented:
> This has deterministic (albeit perhaps surprising if you don't know the implementation) behavior.

Great, thank you for the explanation!

One additional question: will it be possible to use the same infrastructure for stack tracing once we start implementing one of the GC proposals? Seems like that should be the case, but I might well be missing something.

fitzgen commented:
> One additional question: will it be possible to use the same infrastructure for stack tracing once we start implementing one of the GC proposals? Seems like that should be the case, but I might well be missing something.

Yep, doing that should be a lot easier after this lands.

fitzgen commented:
> I assume we can capture the context at the wasm-to-host transition and use `unw_init_local` to initialize the unwind cursor at that context. We could then step the cursor until we encounter a non-wasm frame.

Yes, we could do this. Adding an out-of-line function call (`unw_getcontext`) to every wasm->host transition isn't great though. (On arm, it is actually a macro of some inline assembly, but since Rust doesn't have inline assembly, we would probably wrap that in a C function that we call anways...)

AFAICT, there is no blessed way of re-initializing an existing `unw_context_t` from just a PC and SP, which is all that should really be required for stack walking in principle. On both x86-64 and arm, `unw_context_t` is just a define of a `ucontext_t`, so maybe we could just poke the `ucontext_t`'s register values directly...

peterhuene commented:
Agreed, if we go with such a design, I would want to limit the context capturing only to host functions not defined by Wasmtime (and perhaps *only* via the C API, where there is a chance of another JIT runtime being used).  Perhaps it could even be something embedders opt-in to.

The bottom line is that we shouldn't have to pay the cost for capturing context when calling into Wasmtime's WASI implementation especially, as we are guaranteed to be able to unwind through those frames for both libunwind and Windows.

peterhuene commented:
All that said, I'm comfortable with landing a stack-walking-based GC before we fix the issue that currently only affects .NET hosts.

bjorn3 commented:
Maybe there is some way for the .NET implementation to insert a frame whose unwind info doesn't do normal unwinding, but sets the sp to the value before entering the .NET JITtes code, such that all .NET JITted code is skipped. Or has a special personality function that invokes the .NET unwinder and then sets the register state as the .NET unwinder gives back when unwinding past the JITted code.

tschneidereit commented:
> All that said, I'm comfortable with landing a stack-walking-based GC before we fix the issue that currently only affects .NET hosts (the .NET API doesn't support reference types yet anyway).

While I think that *might* be fine to do, I'm actually quite glad we have this embedding and thus caught this issue.

What's more, even if we find a workaround specifically for .NET, I think it might make sense to take a closer look at whether we can then have sufficient confidence in our ability to make this approach work in all cases that might become relevant. I.e., would landing this begin painting ourselves into a corner that's increasingly hard to get out of at some later time when we realize we need to?

alexcrichton commented:
@fitzgen 

> I'm not sure what you're talking about here. What explicit code generation? What code generation do we do at all that isn't handled by Cranelift?

Oh by "explicit code generation" I mean "the thing that isn't deferred reference counting".

---

So overall I feel like [my thinking above](https://github.com/bytecodealliance/wasmtime/pull/1832#issuecomment-642204938) is still somewhat attractive. That scheme is where on a call to `Store::gc` it doesn't actually do anything if we're missing some blessed marker on the stack (or some other mechanism to detect that our stack trace is likely incomplete).

The downside of that is that runtimes like .NET will GC less, but that seems like it's not really that much of an issue given that our "GC" here is very small. The GC'd aspect is you gave a value to wasm and then it became unreachable during the execution of wasm. If we don't eagerly clean those up it doesn't seem like it's the end of the world, especially because we'll want to eventually fix this.

I think that explicit reference counting is a defunkt strategy given our unwinding strategy. I don't really see how implementing unwinding correctly with explicit reference counting is going to be any less hazardous or any less work than this already was. If that's taken as an assumption then the final appearance of this feature will be exactly this PR except with more changes to the `gc` function and codegen (to manage `JitActivation` entries and such). @fitzgen brings up a good point though that the `backtrace` crate will likely no longer be what we use.

Personally I see a few directions to go from here:

1. Land this PR, work afterwards to fix .NET
2. Don't land this PR, fix .NET, then land this PR. (e.g. `JitActivation` or similar)
3. Switch this PR to explicit reference counting.

I don't think (3) is feasible and I think 1/2 are pretty close. I think it might be good to sketch out in more detail how `JitActivation` could work though.

tschneidereit commented:
> Personally I see a few directions to go from here

I find this convincing, and support (1), given that @peterhuene also [voiced support for it](https://github.com/bytecodealliance/wasmtime/pull/1832#issuecomment-642313755).

fitzgen commented:
Ok so we talked about this in today's wasmtime meeting. The conclusion we came to was to implement a canary stack frame in this PR for now, so we can detect situations where libunwind cannot walk the full stack, and then we can skip GC rather than get potential unsoundness. But long term, we want to do the precise `JitActivation` approach (it would allow other additional benefits, such as getting wasm stack traces for .net traps, which aren't currently available).

Thanks everyone for the discussion and design input!

fitzgen commented:
@alexcrichton ok, I've implemented the stack canary and rebased, so this should be ready for another round of review.

We will probably want a new release of `backtrace` before we merge this (or at least we'll need one before we publish again after merging this).

alexcrichton commented:
Wow that was fast!

I'll take a look tomorrow!

alexcrichton commented on crates/runtime/src/externref.rs:
This might be a bit more compact as `(0..size).map(|_| ...).collect()`

alexcrichton commented on crates/environ/src/vmoffsets.rs:
```diff
@@ -591,6 +607,19 @@ impl VMOffsets {
+/// Offsets for `VMExternRefActivationsTable`.
+impl VMOffsets {
+    /// Return the offset for `VMExternRefActivationsTable::next`.
+    pub fn vm_extern_ref_activation_table_next(&self) -> u32 {
```
I don't actually see this end up getting used anywhere, was this perhaps from a previous iteration?

alexcrichton commented on crates/runtime/src/externref.rs:
```diff
@@ -393,6 +461,704 @@ impl Deref for VMExternRef {
+    ///
+    /// NB: this is an `UnsafeCell` because it is read from and written to by
+    /// compiled Wasm code.
+    next: UnsafeCell<NonNull<TableElem>>,
```
How is this field used from wasm code? I tried to read over but couldn't find much... It also looked like `VMOffsets` contained information about the `end` field below too, but it's not an `UnsafeCell`?

alexcrichton replied:
Ah ok I confirmed on Zulip, but it sounds like this is planned to get used in the future. I presume that the idea is that with `table.get` you'd have a fast path for using these two fields to inline the `try_insert` method, and you'd fallback to an intrinsic for out-of-line growth and such?

Reading over all this dealing with chunks and such feels a bit overly complicated. Could this perhaps be simplified a bit? I'm thinking that this could perhaps be restructured to something that looks like:

```rust
 pub struct VMExternRefActivationsTable {
    // finter into `fast_path_storage`
    next: UnsafeCell<NonNull<TableElem>>,
    // never changes after creation
    end: NonNull<TableElem>,
    // used by `try_insert` and stored via `next` in jit code
    fast_path_storage: Box<[TableElem]>,
    
    imprecise_roots: HashSet<VMExternRef>,
    precise_roots: HashSet<VMExternRef>,
}
```

where here `fast_path_storage` is allocated once and never changes. When an overflow of that table happens it's drained and everything is moved into `imprecise_roots`. That way `imprecise_roots` is a growing set which deduplicates things (in case you `table.get` the same thing a bunch of times) and also handles reallocation for us (we don't need to manually double capacities and such).

On a GC we'd fill in `precise_roots` and then we'd `mem::swap` the precise/imprecise sets and then clear the imprecise one. Maybe with some other trickery around optimizing reference counts or something like that.

I'm mostly hoping that we can simplifiy the `chunks` list because:

* It seems somewhat complicated to manage, especially if we're trying to be clever about capacities and double them
* I'm a bit worried about the segmented-stacks thrashing problem where you keep allocating a doubly-bigger chunk but then freeing it, when it'd be more efficient to just double the size of the internal chunk.
* Roots aren't deduplicated in the chunks so if you pass around the same `VMExternRef` a bunch it may accidentally fill them up quite a lot.

With a persistent hash set I think it'd solve the duplication/capacity issues? Anyway curious what you think about this!

fitzgen replied:
This is a good idea, however

* we may still want to have the fast-path table grow shrink based on the precise size of the last root set (probably fine to skip for now, but we'll likely want to investigate this in the future)

* we lose the ability to reclaim memory once we are no longer using excessive capacity in these sets (again, probably fine for now, but we'll likely want to handle this sometime in the future)

alexcrichton replied:
Switching `fast_path_storage` to a `Vec<T>` would be an easy way to make the fast-path-table growable (that's a good point, and one thing I was curious how often it would come into play). For shrinking we could periodically call `HashSet::shrink_to_fit` I think?

fitzgen replied:
Yeah. Would be nice if `HashSet::shrink_to` was stable, so we could shrink capacity by half when the length is a quarter of the capacity, and keep the amortized O(1). We could do `shrink_to_fit` followed by `reserve` but this is suboptimal.

Anyways, I'm just gonna do the simplest thing here for now, and leave that stuff to when we actually see it in profiles.

alexcrichton commented on crates/runtime/src/externref.rs:
Since this registry only needs to be within a `Store`, how come this is an `RwLock` instead of `RefCell`?

fitzgen replied:
Otherwise `Module` is not `Send`:

```
error[E0277]: `std::cell::RefCell<wasmtime_runtime::externref::StackMapRegistryInner>` cannot be shared between threads safely
   --> crates/wasmtime/src/module.rs:579:5
    |
578 |     fn _assert<T: Send + Sync>() {}
    |                   ---- required by this bound in `module::_assert_send_sync::_assert`
579 |     _assert::<Module>();
    |     ^^^^^^^^^^^^^^^^^ `std::cell::RefCell<wasmtime_runtime::externref::StackMapRegistryInner>` cannot be shared between threads safely
    |
    = help: within `wasmtime_runtime::externref::StackMapRegistry`, the trait `std::marker::Sync` is not implemented for `std::cell::RefCell<wasmtime_runtime::externref::StackMapRegistryInner>`
    = note: required because it appears within the type `wasmtime_runtime::externref::StackMapRegistry`
    = note: required because of the requirements on the impl of `std::marker::Send` for `std::sync::Arc<wasmtime_runtime::externref::StackMapRegistry>`
    = note: required because it appears within the type `wasmtime_runtime::externref::StackMapRegistration`
    = note: required because of the requirements on the impl of `std::marker::Send` for `std::sync::Arc<wasmtime_runtime::externref::StackMapRegistration>`
    = note: required because it appears within the type `std::option::Option<std::sync::Arc<wasmtime_runtime::externref::StackMapRegistration>>`
    = note: required because it appears within the type `std::option::Option<std::option::Option<std::sync::Arc<wasmtime_runtime::externref::StackMapRegistration>>>`
    = note: required because of the requirements on the impl of `std::marker::Send` for `std::sync::Mutex<std::option::Option<std::option::Option<std::sync::Arc<wasmtime_runtime::externref::StackMapRegistration>>>>`
    = note: required because of the requirements on the impl of `std::marker::Send` for `std::sync::Arc<std::sync::Mutex<std::option::Option<std::option::Option<std::sync::Arc<wasmtime_runtime::externref::StackMapRegistration>>>>>`
    = note: required because it appears within the type `module::Module`
```

alexcrichton commented on crates/runtime/src/externref.rs:
Could this assert that `sp` is not null for wasm frames? (help catch bugs in `backtrace`)

alexcrichton commented on crates/runtime/src/externref.rs:
Because of this, could the activation table have a `Drop` which asserts that the `precise_stack_roots` set is empty? Otherwise we'd leak data accidentally b/c there's no dtor which drains the map.

(or maybe we should also assert that all chunks are `None` since they don't get frobbed on `Drop` either?)

fitzgen commented on crates/jit/src/code_memory.rs:
```diff
@@ -134,7 +134,7 @@ impl CodeMemory {
             if !m.is_empty() {
                 unsafe {
-                    region::protect(m.as_mut_ptr(), m.len(), region::Protection::READ_EXECUTE)
+                    region::protect(m.as_mut_ptr(), m.len(), region::Protection::ReadExecute)
```
Yeah, when I get merge conflicts in `Cargo.lock` I generally just blow it away and let the next build recreate it, but this pulled in that problematic dep update we all talked about a couple days ago. Bleh.

fitzgen replied:
Also, the `backtrace` update needs these dep updates.

fitzgen commented on crates/wasmtime/src/func.rs:
```diff
@@ -194,9 +194,17 @@ macro_rules! getters {
+
+                    {
+                        let canary = 0;
+                        let _auto_reset = instance
```
It probably doesn't matter in practice, but that would mean the canary is taken from a stack frame that has since been popped

/me shrugs

I guess I could switch the API to `table.with_stack_canary(|| { ... })` instead of an RAII thing? Would be harder to misuse this way too. I think I'll do that.

alexcrichton replied:
Either way's fine by me, I was mostly just hoping that we can encapsulate everything necessary to enter wasm in one function in wasmtime, which currently is `catch_traps`. I think the `with_stack_canary` call can be embedded in there too? (we can perhaps rename `catch_traps` to `invoke_wasm` or something like that)

fitzgen replied:
Ah I misunderstood what you were asking for in the original comment. This is done now.

fitzgen commented:
> Apart from the inline comments the only other thing I'd say is that I think it'd be good to look into not having two registrations per module (`GlobalFrameInfoRegistration` and `StackMapRegistration`) and trying to lump that all into one (also deduplicating the `BTreeMap` lookups and such)

Ah, I just remembered why it has to be this way: the `GlobalFrameInfo` is in the `wasmtime` crate, but we need the stack map stuff inside the `wasmtime-runtime` crate, so it can't be added as part of `GlobalFrameInfo`. The split between `wasmtime` and `wasmtime-runtime` crates is a bit annoying...

fitzgen commented on crates/runtime/src/instance.rs:
```diff
@@ -72,6 +74,19 @@ pub(crate) struct Instance {
+    ///
+    /// The `vmctx` also holds a raw pointer to the registry and relies on this
+    /// member to keep it alive.
+    pub(crate) stack_map_registry: Arc<StackMapRegistry>,
```
Same question as above regarding `*mut` vs `Arc`: https://github.com/bytecodealliance/wasmtime/pull/1832/files?file-filters%5B%5D=.rs#r439681507

alexcrichton commented on crates/wasmtime/src/instance.rs:
```diff
@@ -183,10 +191,27 @@ impl Instance {
+            imports,
+            host_info,
+            store.externref_activations_table().clone(),
+            store.stack_map_registry().clone(),
```
you can probably avoid passing these two parameters explicitly and read them from the passed-in `store` argument inside the method.

alexcrichton commented on crates/wasmtime/src/module.rs:
```diff
@@ -534,6 +537,41 @@ impl Module {
+    pub(crate) unsafe fn register_stack_maps(
+        &self,
+        registry: &Arc<StackMapRegistry>,
+    ) -> Option<Arc<StackMapRegistration>> {
```
I think that this API may actually be incorrect for GC, because you can register a module into two stores. The second store won't actually get registered in its `registry` which I think means GC won't be precise?

alexcrichton replied:
I think this'll need to look like [`register_jit_code`](https://github.com/bytecodealliance/wasmtime/blob/357fb11f4603afc5f81eb087e00ca9e543ab4db7/crates/wasmtime/src/runtime.rs#L890-L905) ish where we register a compiled module with a `Store` and metadata about that is stored in `Store` rather than the `Module`.

Honestly we should do this with `register_frame_info` too, but I can tackle that some other time.

fitzgen replied:
Good eye! I didn't realize that a single module could be registered with multiple stores!

fitzgen replied:
Okay the new version fixes this by registering with the store, rather than with the module. This also means that we can get rid of `StackMapRegistration` entirely, and can instead rely on the store to keep the registry itself alive!

alexcrichton commented on crates/wasmtime/src/ref.rs:
```diff
@@ -36,6 +36,11 @@ impl ExternRef {
     }
 
+    /// Get the reference count for this `ExternRef`.
+    pub fn get_reference_count(&self) -> usize {
```
Perhaps we should take a leaf out of `Rc`'s book and use [`strong_count`](https://doc.rust-lang.org/std/rc/struct.Rc.html#method.strong_count) for forwards-compatibility with weak references if we ever add them?

alexcrichton commented on crates/wasmtime/src/runtime.rs:
```diff
@@ -194,10 +195,16 @@ impl Config {
-        // The reference types proposal depends on the bulk memory proposal
+
+        self.flags
+            .set("enable_safepoints", if enable { "true" } else { "false" })
```
Out of curiosity, could we always enable this? Presumably once reference types ships we'll enable this unconditionally anyway.

fitzgen replied:
We could, but it implies an additional pass over the IR by cranelift, and if we know there aren't any being used, then we can skip that unnecessary work.

alexcrichton replied:
This seems fine for now, but I wanted to point this out because that optimization is something we need to likely implement in short order in that case. In theory there should be a fast path for functions which don't use anyref?

Put another way, if this is scoped to only reference types for performance reasons, then we need to file an issue about that and get it fixed before fully shipping reference types.

fitzgen replied:
Filed https://github.com/bytecodealliance/wasmtime/issues/1883

alexcrichton commented on tests/all/gc.rs:
```diff
@@ -0,0 +1,228 @@
+        drop_ref.call(&args)?;
+    }
+
+    assert!(num_refs_dropped.get() > 0);
```
Answering my own question, if my thought earlier about the stack canary and rvalue promotion breaks things this assertion will break. If our stack canary *never* works then nothing will get gc'd from the above loop.

fitzgen replied:
Ah, yep, I see you figured it out :)

fitzgen commented on crates/runtime/src/externref.rs:
```diff
@@ -393,6 +429,647 @@ impl Deref for VMExternRef {
+    end: NonNull<TableElem>,
+
+    /// Bump allocation chunk that stores fast-path insertions.
+    chunk: Box<[TableElem]>,
```
Sorry I didn't mention this, but since switching from the list-of-chunks representation to your new suggested one, the table automatically runs the `Drop` implementations, so dropping the table will now properly avoid leaks.

Therefore, I didn't think it was necessary to insert these debug assertions anymore. (Also, inserting them would require that we `gc` before dropping the table, to ensure that it gets swept. Everything is easier when we can rely on automatic `Drop`s!)

alexcrichton replied:
Whoops missed that, nice!

fitzgen commented on crates/wasmtime/src/values.rs:
```diff
@@ -79,15 +79,22 @@ impl Val {
+                let externref_ptr = x.inner.as_raw();
+                if let Err(inner) = store.externref_activations_table().try_insert(x.inner) {
+                    store.gc();
+                    store.externref_activations_table().insert_slow_path(inner);
```
I figured since it is already a slow path, it might as well do de-duplication, rather than use the bump chunk. I can leave a comment to this effect, and also make a method for this idiom.

fitzgen commented on crates/wasmtime/src/func.rs:
```diff
@@ -729,13 +730,18 @@ impl fmt::Debug for Func {
+        let canary = 0;
+        let _auto_reset_canary = store
+            .externref_activations_table()
+            .set_stack_canary(&canary);
```
It does work today (the tests would fail if it didn't, since they are checking the reference counts and asserting that they go down after GC, which means we did sweep the table, which means we saw the canary).

I agree that it is slightly fragile. I don't know what we could do to avoid it without something like `test::blackbox` or calling `getcontext` to get the SP (eww out-of-line call).

fitzgen commented on crates/runtime/src/externref.rs:
Then we would need to always have the `inner: T` member and rely on LLVM to optimize it away when not used in non-`cfg(debug_assertions)`. Does that sound OK to you?

alexcrichton replied:
Yeah given the limited usage of this I trust LLVM enough to figure everything out.

alexcrichton commented on crates/runtime/src/externref.rs:
I think at this point this can be `RefCell`?

alexcrichton commented on crates/wasmtime/src/runtime.rs:
```diff
@@ -792,6 +800,8 @@ pub(crate) struct StoreInner {
     jit_code_ranges: RefCell<Vec<(usize, usize)>>,
     host_info: RefCell<HashMap<HostInfoKey, Rc<RefCell<dyn Any>>>>,
+    externref_activations_table: Rc<VMExternRefActivationsTable>,
+    stack_map_registry: Arc<StackMapRegistry>,
```
I think this can be `RefCell` now?

alexcrichton commented on crates/wasmtime/src/runtime.rs:
```diff
@@ -904,6 +916,40 @@ impl Store {
+            })
+        {
+            // Either there aren't any functions in the module, or we already
+            // registered it in this store.
```
To confirm, is every function guaranteed to have a stack map, even if it doesn't use reference types?

fitzgen replied:
Good eye, this is a bug introduced during all the refactoring today. I'm going to just make `StackMapRegistry::register_stack_maps` idempotent, rather than trying to check if its already registered here. This is a nice little clean up, since the registry is in the best position to answer this question anyways.

alexcrichton commented on crates/wasmtime/src/trampoline/func.rs:
```diff
@@ -210,7 +212,14 @@ pub fn create_handle_with_function(
+        if ft.params().iter().any(|p| *p == ValType::ExternRef)
+            || ft.results().iter().any(|r| *r == ValType::ExternRef)
+        {
+            flag_builder.enable("enable_safepoints").unwrap();
```
Oh dear this is nasty, can we load flags from the `Store` instead of trying to recreate them here though?

alexcrichton replied:
(in that I think we have a `Box<dyn TargetIsa>` shoved somewhere in there I think

alexcrichton commented on crates/wasmtime/src/types.rs:
```diff
@@ -106,6 +106,7 @@ impl ValType {
             ValType::F32 => Some(ir::types::F32),
             ValType::F64 => Some(ir::types::F64),
             ValType::V128 => Some(ir::types::I8X16),
+            ValType::ExternRef => Some(ir::types::R64),
```
Mind throwing a `cfg!` here for the target pointer width? (and below too)

fitzgen commented:
I had to `cfg(target_arch = "x86_64")` the reference types-related tests because Cranelift doesn't support safepoints/stack maps for any other targets (notably aarch64).

fitzgen commented:
Confused why the `aarch64` CI job is [running the reference types tests](https://github.com/bytecodealliance/wasmtime/pull/1832/checks?check_run_id=774637548#step:10:1199) despite them [being ignored if `!cfg!(target_arch = "x86_64")`](https://github.com/bytecodealliance/wasmtime/pull/1832/commits/6a88d5cbec45e325bb22900f886b448bf4a80b8a#diff-a7b0a2dee0126cddf994326e705a91eaR215). Feel like I must be missing some simple typo or something like that, but I just don't see it...

fitzgen commented:
Gah, its because the `build.rs` is being compiled for `x64`.

alexcrichton commented on tests/all/main.rs:
```diff
@@ -18,3 +18,7 @@ mod table;
 mod wast;
+
+// Cranelift only supports reference types on x64.
+#[cfg(target_arch = "x86_64")]
```
For future instnaces of this, mind tagging this with a FIXME and an issue number so when we get around to aarch64 reference types we can make sure we run all the tests?

alexcrichton commented on build.rs:
```diff
@@ -209,9 +209,14 @@ fn ignore(testsuite: &str, testname: &str, strategy: &str) -> bool {
+            | ("reference_types", "externref_id_function") => {
+                // Ignore if this isn't x64, because Cranelift only supports
+                // reference types on x64.
+                return env::var("CARGO_CFG_TARGET_ARCH").unwrap() != "x86_64";
```
Same comment here for the aarch64 testing
